[WIP] MultiField_Stats Aggregation

nknize opened the pull request:
re: issue #16817 this is a WIP PR (for initial feedback and review) to add the first Multi-Field metric agg that computes the Pearson product-moment correlation coefficient for a given list of numeric fields.

It's a pretty simple single pass calculation that I simulated in Matlab validating against Matlab's `corr` method. This PR includes a validation of the actual aggregation results (computed across shards) against the expected value computed using a large single set (see CorrelationIT). Initial rough calculations over a set of 1M distributed observations give at least 1e-7 accuracy.

UPDATE: This aggregation is now called `multifield_stats`. It takes a list of multiple fields and computes the following statistics:
1. counts
2. means
3. variances
4. skewness
5. kurtosis
6. covariance
7. correlation

wamuir commented:
Define correlation matrix for x1...x4:
`corr = numpy.array(`
`[[1.00, .397, .214, .569],`
` [.397, 1.00, .448, .439],`
` [.214, .448, 1.00, .404],`
` [.569, .439, .404, 1.00]])`

Generate some data (10M obs):
`numpy.random.seed(seed=1)`
`mtx = numpy.random.randn(4, 10000000)`
`chol = numpy.linalg.cholesky(corr)`
`data = numpy.dot(chol, mtx)`

`numpy.set_printoptions(precision=17)`
`print(numpy.corrcoef(data))`
`[[ 1.                   0.39650687829250175  0.21417973223669709
   0.56896721641827785]`
 `[ 0.39650687829250175  1.                   0.44797611026090212
   0.43881323671321049]`
 `[ 0.21417973223669709  0.44797611026090212  1.                   0.40411226138655432]`
 `[ 0.56896721641827785  0.43881323671321049  0.40411226138655432  1.                 ]]`

numpy.matrix.transpose(data) -> index as doubles -> es query: "pearson_correlation":
`[[ 1.                                 ]`
` [ 0.3965068782924764  1.                                             ]`
` [ 0.21417973223672318  0.44797611026089856   1.          ]`
` [ 0.5689672164183015  0.43881323671322076  0.40411226138653933  1.        ]]`

Very good

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/correlation/Correlation.java:
```diff
@@ -0,0 +1,31 @@
+ * Interface for Correlation Metric Aggregation
+ */
+public interface Correlation extends Aggregation {
+    long count();
```
Could you add Javadocs to these methods?

colings86 commented on core/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceAggregatorBuilder.java:
```diff
@@ -0,0 +1,465 @@
+        }
+        for (Script script : scripts) {
+            ValuesSourceConfig<VS> config = config(context, null, script);
+            configs.put(script.getType().getParseField().getPreferredName(), config);
```
This name won't work as I can specify multiple scripts of a single type (e.g. `inline`) and the `ParseField` name for them all will be `"inline"` so they will overwrite each other. I think we need to go back to the parser here and parse a `Map<String, Script>` so each script can have a name. In the request this would look like:

```
{
    ...
    "script": {
        "my_script_1": {
            "inline": "script contents",
            "lang": "expressions"
        },
        "my_script_2": "script contents",
        ...
    }
```
Note that scripts can either be a JSON object or a String. The `Script.parse()` method handles both cases.

colings86 commented:
@nknize I left a couple of comments but I like it so far

wamuir commented:
Consider implementing pairwise deletion and making this the default.  For one potential gotcha w/ listwise deletion, using the example provided by @nknize in #16817, consider replacing `FOO_Corp_Stock_Value` with `FB` or `TWTR`:

```javascript
"pearson_correlation" : {
    "correlation" : {
        "field" : [ "FOO_Corp_Stock_Value", "AAPL", "INDEXSP" ],
    }
}
```

Also consider if the ticker symbol changed for "FOO_Corp_Stock" during this period, and the tickers are indexed separately.  This throws an exception, but the best we could get back with listwise deletion is null output.  More importantly, consider if data aren't missing at random and missingness on A moderates the relationship beween B and C.


Also related, @polyfractal mentions tests for correlation in #16817:

> Time-series data isn't i.i.d. (current value is usually related to last value in some fashion), so you'd likely need other pipeline aggs to first remove inter-series dependence before **testing for correlation**

If this agg cannot accept sub-aggregations (e.g., extended stats) then consider returning the appropriate stats with the agg.  I'd argue that the appropriate stats are extended stats (incl. counts, means and variance) for variables, by pair.  Combined with an option for listwise deletion, one should be able to produce a variance-covariance matrix and perform many statistical tests.

@polyfractal also mentions serial correlation.  Down the road, there probably should be a way to examine residuals if this agg is extended to test hypotheses about a population beyond the bivariate case.  But the ouput now seems sound: unless I am mistaken, the sample correlation coefficient alone (as output now) should be an unbiased estimate of the population parameter to the extent that the items are normally distributed and were measured without error; it is the standard error of this estimate (and therefore test statistics) that is impacted.  Now, this isn't to say that the coefficient won't change after partialling out one or more lags, but that might be best viewed as a specification issue on the part of the user.

For example, in the case of OLS (which would be a nice implementation, down the road), the test statistic makes distributional assumptions of the residual such that all elements along the diagonal of the residual var/cov matrix equal a constant (homogeneity) and elements off the diagonal are zero (independence).  In the case that this i.i.d. assumption is violated, beta estimates are still unbiased and consistent but variance estimates of OLS estimators are both biased and inconsistent---standard errors are not trustworthy.

nknize commented:
This is some wonderful feedback! I've had a chance to revisit this a bit and should have an updated commit in a day or so. 

> Consider implementing pairwise deletion and making this the default.

Will be addressing this shortly. Though I think there's plenty of time to get this as an enhancement in another PR?

> consider if the ticker symbol changed for "FOO_Corp_Stock" during this period

Here we're assuming the stock ticker is the fieldname. So if the ticker symbol changes the application will need to map that to the field name. 

> consider if data aren't missing at random and missingness on A moderates the relationship beween B and C.

I think this is a biggie. I'm looking at addressing this by way of count based weights. But I'm just getting started on this part. Again, might hold off to another PR?

Based on the feedback I'm thinking of making the following changes:

1. The agg will be changed to be more than just correlation. So I'll be refactoring to `multifield_stats`. Any objection to the name? The `multifield_stats` will provide
    * `count` - per field count to (eventually) communicate missing values
    * `mean` - per field means
    * `covariance` - upper triangle matrix (to include diagonals for variance) of the variance-covariance
2. In addition to the above, a sister `extended_muitifield_stats` (not really liking the name so I'm open to suggestions) will provide the following:
    * `correlation` - upper triangle matrix (to include diagonals for autocorrelation) of the pearson product-moment correlation coefficients
    * `kurtosis` - scalar value describing the shape of the distribution
    * `skewness` - scalar measurement describing the symmetry of the distribution about the mean

I've completed counts, mean, covariance, and correlation. Before I split into 2 separate aggs and add kurtosis and skewness I wanted to solicit feedback for the usefulness of the additional measurements and for having 2 separate aggregations.

nknize commented:
Updated the PR to do the following:

* rename aggregation to `multifield_stats`
* adds `mean`, `variance`, `skewness`, `kurtosis`, `covariance`
* updates `singleFieldTest` to include testing all generated statistics

Still todo in this PR:
* add testing for missing and multi values
* incorporate @colings86 feedback re: script names
* update javadocs

nknize commented:
@colings86 If you get a chance can you have a look at commit 61db3e8? Make sure I didn't miss anything while trying to parse the Scripts as a Map.

nknize commented:
So the PR has gotten pretty big, but everything should be ready. This does not yet handle pairwise missing values. As discussed above that can be handled in a separate PR.

Here is an example of the output:

```javascript
  "aggregations" : {
    "multifieldstats" : {
      "counts" : {
        "randVal1" : 100000,
        "value" : 100000,
        "randVal2" : 100000,
        "values" : 100000
      },
      "means" : {
        "randVal1" : 0.5007814114859065,
        "value" : 50000.499999999134,
        "randVal2" : 0.5004831083181183,
        "values" : 50001.999999999134
      },
      "variances" : {
        "randVal1" : 0.08361899306430848,
        "value" : 8.333416666666644E8,
        "randVal2" : 0.0832774520285396,
        "values" : 8.333416666666644E8
      },
      "skewness" : {
        "randVal1" : -0.0030206367840570187,
        "value" : 7.154127608560924E-17,
        "randVal2" : -7.692826885603065E-4,
        "values" : 7.154127608560924E-17
      },
      "kurtosis" : {
        "randVal1" : 1.8005677156780868,
        "value" : 1.799999999760016,
        "randVal2" : 1.8008100300780934,
        "values" : 1.799999999760016
      },
      "covariance" : {
        "randVal1" : {
          "value" : 58.078219752650746,
          "randVal2" : -5.485860518247862E-4,
          "values" : 58.078219752650746
        },
        "value" : {
          "randVal2" : -7.158019911939834,
          "values" : 8.333416666666662E8
        },
        "randVal2" : {
          "values" : -7.158019911939815
        }
      },
      "correlation" : {
        "randVal1" : {
          "value" : 0.006957436968003296,
          "randVal2" : -0.006573983049459838,
          "values" : 0.006957436968003296
        },
        "value" : {
          "randVal2" : -8.59246237772021E-4,
          "values" : 1.000000000000002
        },
        "randVal2" : {
          "values" : -8.592462377720186E-4
        }
      }
    }
  }
```

It might be nice (in another PR) to specify which stats the user would like returned. It may not change what is computed, but simplifies the output if the application only cares about certain fields.

/cc @polyfractal @colings86 @wamuir

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/InternalMultiFieldStats.java:
```diff
@@ -0,0 +1,254 @@
+    @Override
+    public double getMean(String field) {
+        if (multiFieldStatsResults == null || multiFieldStatsResults.means.containsKey(field) == false) {
+            return Double.NaN;
```
General thought for these various `Double.NaN`'s .. should we be returning `null` instead (and subsequently returning a `Double` instead of `double`)?  E.g. because some fields may actually have `NaN`'s as real values, and lacking a field isn't quite the same as not-a-number?

Not sure, probably needs several more opinions :)

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/InternalMultiFieldStats.java:
```diff
@@ -0,0 +1,254 @@
+        aggregations.removeIf(p -> ((InternalMultiFieldStats)p).multiFieldStatsResults == null);
+
+        // return empty result iff all stats are null
+        if (aggregations.size() == 0) {
```
Maybe an `isEmpty()` here?

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsAggregator.java:
```diff
@@ -0,0 +1,147 @@
+                    // get the field value (multi-value is the average of all the values)
+                    double fieldValue = 0;
+                    for (int i = 0; i < valuesCount; ++i) {
+                        fieldValue += doubleValues.valueAt(i);
```
Should we check/ignore `NaN` here?  A single `NaN` will "infect" the average and turn it all into `NaN`, which is technically correct...but also not very useful.

I don't have a strong opinion, just wanted to raise the idea.

nknize replied:
++ Nice catch! I should've had that in there.

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsAggregatorBuilder.java:
```diff
@@ -0,0 +1,75 @@
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.aggregations.AggregatorFactories;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.support.*;
```
Wildcard imports will break the check-style build (ask me how I know :D)

nknize replied:
argh! I need to fix my IDE setting. Thx!

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java:
```diff
@@ -0,0 +1,431 @@
+ *
+ * @internal
+ */
+class MultiFieldStatsResults implements Streamable {
```
Could the calculation of stats and serialization of the results be separated?  Then the Results class could be Streamable, final (no need for dirty flag) and deal with merging code, while the computation just deals with adding docs to the stats and doesn't have to special case the serialization code?

Might make it easier to unit-test just the calculation portion too.

Not sure, I'd seek a second opinion :)

nknize replied:
Yes! I went back and forth including that in this PR (even had it in a todo). I think it makes sense doing this before committing so I'll make the change.

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java:
```diff
@@ -0,0 +1,431 @@
+    /** kurtosis values */
+    protected HashMap<String, Double> kurtosis;
+    /** covariance values */
+    protected HashMap<String, HashMap<String, Double>> covariances;
```
This essentially represents the upper triangle covariance matrix, right?  I wonder if future iterations on this could use a flat vector of values and look up values based on an offset.  I suppose it would also need a LUT to convert field names to row/col positions too.

nknize replied:
I thought about something like that. For initial dev I chose to keep it simple. :) If we think it will buy anything we can make the change later?

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java:
```diff
@@ -0,0 +1,431 @@
+    public double getCovariance(String fieldX, String fieldY) {
+        if (fieldX.equals(fieldY)) {
+            return variances.get(fieldX);
+        } else if (variances.containsKey(fieldX) == false && variances.containsKey(fieldY) == false) {
```
Should this be an OR?  E.g. if either of the fields are missing, rather than both?  Looks like it could cause problems if one field is missing when it gets to `getValFromUpperTriangularMatrix()`?

nknize replied:
No, but the way this is currently written is certainly trappy. I'll fix and add comments so its (hopefully) more clear.

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java:
```diff
@@ -0,0 +1,431 @@
+            if (map.get(fieldX).containsKey(fieldY)) {
+                return map.get(fieldX).get(fieldY);
+            } else {
+                return map.get(fieldY).get(fieldX);
```
e.g. from above comment, if `fieldX` exists, `fieldX` does not contain `fieldY`, and `fieldY` does not exist... it'll blowup.

But perhaps that's an impossible situation for this code to get into?

nknize replied:
Related to above. I'll fix the trappiness.

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java:
```diff
@@ -0,0 +1,431 @@
+        }
+
+        // deep copy of hash keys (field names)
+        ArrayList<String> cFieldNames = new ArrayList(docFields.keySet());
```
`ArrayList<>(docFields.keySet())` just so there isn't an unchecked warning

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsResults.java:
```diff
@@ -0,0 +1,431 @@
+        // update cache flag
+        dirty = true;
+        final double countA = docCount;
+        final double countB = other.docCount;
```
Should we need a check first to make sure `other` isn't null?  I know the nulls are removed in `doReduce()` above, but thinking about future-proofing in case code gets re-arranged?

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/support/AbstractMultiValuesSourceParser.java:
```diff
@@ -0,0 +1,275 @@
+import org.joda.time.DateTimeZone;
+
+import java.io.IOException;
+import java.util.*;
```
Wildcard import

polyfractal commented on core/src/main/java/org/elasticsearch/search/aggregations/support/MultiValuesSourceAggregatorBuilder.java:
```diff
@@ -0,0 +1,466 @@
+import org.joda.time.DateTimeZone;
+
+import java.io.IOException;
+import java.util.*;
```
Wildcard import

polyfractal commented on core/src/test/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsIT.java:
```diff
@@ -0,0 +1,380 @@
+                .field(valFieldNames[0], vals.get(valFieldNames[0]))
+                .field(valFieldNames[1], vals.get(valFieldNames[1]))
+                .field(valFieldNames[2], vals.get(valFieldNames[2]))
+                .startArray(valFieldNames[3]).value(i+2).value(i+3).endArray()
```
Perhaps add a field that sporadically has `null` or `NaN`, just to make sure sparse data doesn't catastrophically break things?

nknize replied:
hmmm... I can add random field omission? Since these are numeric fields an IAE is thrown if `null` or `NaN` is passed.

polyfractal commented on core/src/test/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/MultiFieldStatsIT.java:
```diff
@@ -0,0 +1,380 @@
+            // check fieldCount
+            assertThat(multiFieldStats.getFieldCount(fieldName), equalTo(multiFieldStatsResults.counts.get(fieldName)));
+            // check mean
+            assertThat(multiFieldStats.getMean(fieldName), closeTo(multiFieldStatsResults.means.get(fieldName), TOLERANCE));
```
Consider `org.elasticsearch.test.hamcrest.DoubleMatcher.nearlyEqual()` instead of `closeTo()`.  Absolute tolerances tend to break when the values are very large (returns false due to floating point rounding error) or when the values are very small (returns true, since the absolute difference between two very small numbers is also very small)

https://github.com/elastic/elasticsearch/blob/master/core/src/test/java/org/elasticsearch/test/hamcrest/DoubleMatcher.java

polyfractal commented:
@nknize Left some comments, I think it's looking good!  I think my main concern is `MultiFieldStatsResults` being mutable ... that dirty flag scares me :)  It might be cleaner split into a class that accumulates and calculates the per-shard results, then converts itself into an immutable result class which is serialized / publicly consumed?  I'd request a second opinion from someone more experienced though :)

Otherwise, just general questions around handling missing values and various `NaN`'s.  Don't have a strong opinion about those, but it's something that I found chronically affecting the pipeline work we did, so wanted to raise the issues.

Future thought:  it might be nice if there was an option to control how multi-values are accumulated per-field (e.g. min/max/sum/avg, etc).  But definitely not important now, this PR is big enough :)

polyfractal commented:
Oh, also, probably needs some reference docs?

nknize commented:
Updated the PR based on feedback. I also changed the output to list the stats by field. I think this will be easier to parse than returning the output by stat name. The following is an example:

```javascript
{
  "aggregations" : {
    "multifieldstats" : {
      "randVal1" : {
        "count" : 5006,
        "mean" : 0.5037337010204279,
        "variance" : 0.08204460149826266,
        "skewness" : -0.029959224026464074,
        "kurtosis" : 1.8104796506497558,
        "covariance_correlation" : {
          "value" : {
            "covariance" : 9.787449451489842,
            "correlation" : 0.011865786992191823
          },
          "randVal2" : {
            "covariance" : -3.6257133839528744E-4,
            "correlation" : -0.00442761976612947
          }
        }
      },
      "value" : {
        "count" : 5006,
        "mean" : 4983.837594886138,
        "variance" : 8292700.078114813,
        "skewness" : 0.02045413721293637,
        "kurtosis" : 1.8080183355015502,
        "covariance_correlation" : {
          "randVal1" : {
            "covariance" : 9.787449451489842,
            "correlation" : 0.011865786992191823
          },
          "randVal2" : {
            "covariance" : 6.466702461177619,
            "correlation" : 0.00785482932207694
          }
        }
      },
      "randVal2" : {
        "count" : 5006,
        "mean" : 0.4964204358056165,
        "variance" : 0.08173278440418814,
        "skewness" : 0.014828068008907269,
        "kurtosis" : 1.8324182586996796,
        "covariance_correlation" : {
          "randVal1" : {
            "covariance" : -3.6257133839528744E-4,
            "correlation" : -0.00442761976612947
          },
          "value" : {
            "covariance" : 6.466702461177619,
            "correlation" : 0.00785482932207694
          }
        }
      }
    }
  }
}

```

nknize commented:
/cc @rashidkpc @spalger The output format for the `multifield_stats` aggregation is up in the air. Would love your feedback from the Kibana Viz perspective.

rashidkpc commented:
@nknize Have you seen this sort of result visualized somewhere before? I'm trying to figure out how this data would be best presented. I suspect that our existing charts wouldn't do it justice.

nknize commented:
@rashidkpc I spoke with @spalger a little bit about this at the all hands. The tabular heatmap visualization is typical for a correlation matrix. Strong positive correlation (closer to +1.0) is usually depicted as a red cell with weaker correlation (closer to 0.0) colored white and strong negative correlation (closer to -1.0) colored blue.

Here's a good example:

![network_correlations](https://cloud.githubusercontent.com/assets/830187/13829864/42daf0aa-eb97-11e5-9776-7ba4647340c3.png)

spalger commented:
Yeah, I'm fine with that structure. We'll have to transform it from object to array, but as long as the fields aren't in a meaningful order this is fine.

rashidkpc commented:
@nknize Can you add a simplified example of the current request and response format?

nknize commented:
@rashidkpc Sure thing. Here's a simple 2 variable example using the same data as above:
* poverty
* income

Request:
```javascript
"aggs": {
    "multifieldstats": {
        "multifield_stats": {
            "field": ["poverty", "income"]
        }
     }
}
```

Response:
```javascript
"aggregations": {
    "multifieldstats": {
        "income": {
            "count": 50,
            "mean": 51985.1,
            "variance": 7.383377037755103E7,
            "skewness": 0.5595114003506483,
            "kurtosis": 2.5692365287787124,
            "covariance_correlation": {
                "poverty": {
                    "covariance": -21093.65836734694,
                    "correlation": -0.8352655256272504
                }
            }
        },
        "poverty": {
            "count": 50,
            "mean": 12.732000000000001,
            "variance": 8.637730612244896,
            "skewness": 0.4516049811903419,
            "kurtosis": 2.8615929677997767,
            "covariance_correlation": {
                "income": {
                    "covariance": -21093.65836734694,
                    "correlation": -0.8352655256272504
                }
            }
        }
    }
}
```

nknize commented:
@clintongormley how about making this feature experimental for 5.0.0?

brwe commented on core/src/main/java/org/elasticsearch/search/aggregations/metrics/stats/multifield/RunningStats.java:
```diff
@@ -0,0 +1,312 @@
+            delta.put(fieldName, fieldValue * docCount - fieldSum.get(fieldName));
+
+            // update running mean, variance, skewness, kurtosis
+            if (means.containsKey(fieldName) == true) {
```
Could we add a reference here? Some paper, book chapter, wikipedia article or something.

clintongormley commented:
Superseded by https://github.com/elastic/elasticsearch/pull/18300
